auslib/db.py

```diff
@@ -1755,21 +1758,44 @@ def __init__(self, db, metadata, dialect):
         self.table.append_column(Column('data', BlobColumn(dataType), nullable=False))
         AUSTable.__init__(self, db, dialect, scheduled_changes=True, scheduled_changes_kwargs={"conditions": ["time"]})
 
-    def getPotentialRequiredSignoffs(self, affected_rows, transaction=None):
-        potential_required_signoffs = []
+    def getPotentialRequiredSignoffs(self, affected_rows, transaction=None, batch=False):
+        potential_required_signoffs = {}
+        rows = []
         for row in affected_rows:
             if not row:
                 continue
-            # Releases do not affect live updates on their own, only the
-            # product+channel combinations specified in Rules that point
-            # to them. We need to find these Rules, and then return _their_
-            # Required Signoffs.
-            info = self.getReleaseInfo(name=row["name"], transaction=transaction)
-            if info:
-                info = info[0]
-                for rule_id in info["rule_ids"]:
-                    rule = self.db.rules.select(where=[self.db.rules.rule_id == rule_id], transaction=transaction)[0]
-                    potential_required_signoffs.extend(self.db.rules.getPotentialRequiredSignoffs([rule], transaction=transaction))
+            rows.append(row)
+        info = self.getReleaseInfo(name=[row['name'] for row in rows] if batch else rows[0]['name'], transaction=transaction, batch=batch)
+        # Releases do not affect live updates on their own, only the
+        # product+channel combinations specified in Rules that point
+        # to them. We need to find these Rules, and then return _their_
+        # Required Signoffs.
+        if info:
+            # get all rules as a union
+            q_rules = [[self.db.rules.rule_id == rule_id] for row in info for rule_id in row['rule_ids']]
+            all_rules = self.db.rules.select(where=q_rules, transaction=transaction)
+
+            # map rules according to rule_id
+            rules_map = {rule['rule_id']: rule for rule in all_rules}
+
+            rs_cache = {}
+
+            for row in info:
```

Comment:
In my local testing, this block still took about 75% of the execution time (about 600ms). In production, this is likely to mean page load times are still over 10 seconds. Would you mind trying to speed up Rules.getPotentialRequiredSignoffs, too? That can be a done in a follow-up pull request, if you prefer.

Reply:
Yeah, I could try fixing this in this PR though I would be more comfortable doing it in a follow-up pull request. Perhaps we could create/file an issue/bug which depends on this PR then have it assigned to me. I could then tackle it once this is merged.

Reply:
Yup, that sounds perfect. If you have access to file bugs, feel free to file it. (If you don't, we should fix that!)